src/random.h

```diff
@@ -67,7 +67,17 @@
  * Thread-safe.
  */
 void GetRandBytes(unsigned char* buf, int num) noexcept;
+/** Generate a random integer in the range [0..range). Precondition: range > 0 */
 uint64_t GetRand(uint64_t nMax) noexcept;
+// The following GetRand*(std::chrono::*) MUST NOT be templated on the type of
+// the duration to prevent unsafe use at the call site. If it was templated, it
+// would be impossible for the compiler to select the right template. E.g.
```

Comment:
is it impossible, or does the compiler just select template duration of param thats passed & then tries to convert afterwards? 

seems that way to me because `std::chrono::seconds val = GetRandDur(std::chrono::hours)` gives a type conversion error

Reply:
No it doesn't give a type conversion error.

This compiles fine for me and the test pass:

```cpp
template <typename D>
D GetRandDur(const D& duration_max)
{
    return D{GetRand(duration_max.count())};
}

BOOST_AUTO_TEST_CASE(util_time_GetRandTime)
{
    std::chrono::seconds rand_hour = GetRandDur(std::chrono::hours{1});
    // Want seconds to be in range [0..1hour), but always get zero :((((
    BOOST_CHECK_EQUAL(rand_hour.count(), 0);
}

Reply:
oh interesting. maybe its different in gcc vs clang?? or I'm missing something obvious.

Reply:
I hope not; that would violate the language spec.

std::chrono::hours should always be implicitly convertible into std::chrono::seconds.

Reply:
ok I tried again and it works. I must have made an accidental error. my mistake!

Reply:
oh, my error was in what I was typing here. I get conversion issue when I do `std::chrono::minutes val = GetRandDur(std::chrono::seconds{120})` 

so when function arg is higher precision than return value. 

which is why I was suspecting compiler selects template duration based on function arg, then tries to convert afterwards. so selection is deterministic vs unexpected, but can lead to unexpected results.